gaphor/UML/actions/partition.py

```diff
@@ -1,154 +1,139 @@
+        self.watch("num_partitions", self.update_shapes)
 
-    @property
-    def toplevel(self):
-        return self._toplevel
-
-    def pre_update(self, context):
-        assert self.canvas
-
-        self._header_size = self.shape.size(context)
-
-        # get subpartitions
-        children: List[PartitionItem] = [
-            k for k in self.canvas.get_children(self) if isinstance(k, PartitionItem)
-        ]
-
-        self._toplevel = self.canvas.get_parent(self) is None
-        self._subpart = len(children) > 0
-        self._bottom = not (self._toplevel or self._subpart)
-
-        if self._toplevel:
-            self._header_size = self._header_size[0], self.DELTA
-
-        handles = self.handles()
+    num_partitions: attribute[int] = attribute(
+        "num_partitions", int, default=INIT_NUM_PARTITIONS
+    )
 
-        # toplevel partition controls the height
-        # partitions at the very bottom control the width
-        # middle partitions control nothing
-        for h in handles:
-            h.movable = False
-            h.visible = False
-        if self._bottom:
-            h = handles[1]
-            h.visible = h.movable = True
-        if self._toplevel:
-            h1, h2 = handles[2:4]
-            h1.visible = h1.movable = True
-            h2.visible = h2.movable = True
+    partition = association("partition", UML.ActivityPartition, composite=True)
 
-        if self._subpart:
-            wsum: int = sum(sl.width for sl in children)
-            self._hdmax = max(sl._header_size[1] for sl in children)
+    def pre_update(self, context: DrawContext) -> None:
+        """Set the min width of all the swimlanes."""
+        self.min_width = 150 * self.num_partitions
 
-            # extend width of swimline due the children but keep the height
-            # untouched
-            self.width = wsum
+    def post_update(self, context: DrawContext) -> None:
+        """Nest the Activity Partitions under the Activity.
 
-            dp = 0
-            for sl in self.canvas.get_children(self):
-                x, y = sl.matrix[4], sl.matrix[5]
-
-                x = dp - x
-                y = -y + self._header_size[1] + self._hdmax - sl._header_size[1]
-                sl.matrix.translate(x, y)
-
-                sl.height = sl.min_height = max(0, self.height - self._header_size[1])
-                dp += sl.width
+        This needs to be done after the update because the activity's
+        classifier doesn't always exist yet.
+        """
+        activity = self.diagram.namespace
+        for partition in self.partition:
+            if (
+                not partition.activity
+                and activity.ownedClassifier
+                and isinstance(activity.ownedClassifier[0], UML.Activity)
+            ):
+                partition.activity = activity.ownedClassifier[0]
+
+    def update_partitions(self) -> None:
+        """Add and remove UML.ActivityPartitions."""
```

Comment:
I think this logic can move to partitionPage.

Reply:
Ok, I'll look in to refactoring this to the partitionpage, thanks!